Network policy latency test clients v.0.0.1

dlapcevic opened the pull request:
A golang client application that runs in pods inside a K8s cluster, and measures network policy enforcement latency for changes to pods and network policies. 

Details are in the README.md file.

Main purpose: Test network policy performance and scalability by using it with [ClusterLoader2](https://github.com/kubernetes/perf-tests/tree/master/clusterloader2) framework. 

/kind feature

dlapcevic commented:
/assign @alan-kut

alan-kut commented on network/tools/network-policy-enforcement-latency/pkg/client.go:
```diff
@@ -0,0 +1,545 @@
+		}
+	}(c.metricsServer, context.TODO())
+
+	if c.config.testPodCreation {
+		if err := c.measurePodCreation(); err != nil {
+			klog.Errorf("Pod creation test failed, error: %v", err)
+		}
+	} else {
+		if err := c.measureNetPolicyCreation(); err != nil {
+			klog.Errorf("Pod creation test failed, error: %v", err)
+		}
+	}
```
testPodCreation is vague.
Maybe it would be fine if this is additional test you can run, e.g.

if testPodCreation {
   c.measurePodCreation()
}

but in the current form it is not really a boolean whether to test pod creation but rather a flag for picking a mode what to test.

Why can't we just test both of them?
What is the usecase for this flag?

dlapcevic replied:
I changed the flag name to `podCreationMode`.

I think it’s better to have one test client to run just one of the tests (modes), because of the use case and to keep it simpler.
You can run both at the same time by deploying them in separate pods for each mode.

The flag determines which mode to use. They work differently.
Default is network policy enforcement latency for network policy creation. It lists target pods and tries to reach them (no pod watcher).
When the flag is `True`, it tests network policy enforcement latency for pod creation (creates pod watcher).

dlapcevic replied:
Split into two applications.

alan-kut commented on network/tools/network-policy-enforcement-latency/pkg/client.go:
```diff
@@ -0,0 +1,545 @@
+// network policy creation.
+func (c *TestClient) measureNetPolicyCreation() error {
+	klog.Infof("Starting to measure network policy enforcement latency on network policy creation")
+	podList, err := c.targetPodList()
```
What about getTargetPodList() or listTargetPods()?

dlapcevic replied:
Changed name to `listTargetPods()`.

alan-kut commented on network/tools/network-policy-enforcement-latency/pkg/client.go:
```diff
@@ -0,0 +1,545 @@
+		return nil, fmt.Errorf("failed to list pods for selector %q in namespace %q, error: %v", c.config.targetLabelSelector, c.config.targetNamespace, err)
+	}
+
+	if len(podList.Items) == 0 {
```
Shouldn't it be equal to expectedTargets?

dlapcevic replied:
Changed the flag name to `maxTargets`, because that's what it actually does.
There will be an error if there are no target pods -- the test cannot carry on.
It will not fail if there are more than `maxTargets`, it will just limit the kube API list call to that value.

alan-kut replied:
what if there is just one pod returned?

dlapcevic replied:
It will run the test for one pod. Send requests to it until it's reached, and generate one latency metric data point for it.

alan-kut commented on network/tools/network-policy-enforcement-latency/pkg/client.go:
```diff
@@ -0,0 +1,548 @@
+	}
+	klog.Infof("%d pods listed", len(podList))
+
+	c.policyCreatedTime = &timeWithLock{lock: sync.RWMutex{}}
```
Why is it intialized here, not in the constructor?

dlapcevic replied:
The test client runs in one of the modes, and it instantiates objects for the specific mode once that mode is starting to run.
I think it's better to instantiate it closer to use, rather than doing it early for all variables, even the ones not to be used.

dlapcevic replied:
Fixed

alan-kut commented on network/tools/network-policy-enforcement-latency/pkg/client.go:
```diff
@@ -0,0 +1,548 @@
+			failed = true
+		}
+
+		policyCreateTime = networkPolicy.GetCreationTimestamp().Time
+		c.policyCreatedTime.time = &policyCreateTime
```
This is still executed, even if the get failed.

dlapcevic replied:
Fixed.

alan-kut commented on network/tools/network-policy-enforcement-latency/pkg/client.go:
```diff
@@ -0,0 +1,548 @@
+
+// reportReachedTimeForPolicyCreation records network policy enforcement latency
+// metric for network policy creation.
+func (c *TestClient) reportReachedTimeForPolicyCreation(target *targetSpec, reachedTime time.Time) {
```
Why not to return errors?

dlapcevic replied:
I don't see the benefit of returning an error here. Logging a warning in case it fails is enough.
It's used in `recordFirstSuccessfulRequest()`, which is called many times as a separate go routine and won't impact the main execution.

alan-kut commented on network/tools/network-policy-enforcement-latency/pkg/client.go:
```diff
@@ -0,0 +1,548 @@
+		policyCreateTime = networkPolicy.GetCreationTimestamp().Time
+		c.policyCreatedTime.time = &policyCreateTime
+	} else {
+		policyCreateTime = *c.policyCreatedTime.time
```
Do you expect this method to be called multiple times? It seems it does nothing after first call.

dlapcevic replied:
Yes. It will be called every time a target is reached. To avoid getting the network policy from apiserver every time, I cache it the first time it’s retrieved.

After the first call, it will reuse the cached `c.policyCreateTime.time`.

alan-kut commented on network/tools/network-policy-enforcement-latency/pkg/client.go:
```diff
@@ -0,0 +1,548 @@
+		default:
+		}
+
+		_, err := utils.RunCommand(command)
+		if err == nil {
```
if err != nil {
  continue + some comment why it's expected
}
the code if err is not nil?

dlapcevic replied:
Done.
Added `err != nil {continue}` and a comment for it.
“CURL command returns errors for requests denied by a policy, or when the IP is unreachable. Everything else is considered a success”

alan-kut reviewed (changes requested):
I haven't reviewed it fully yet.
Here is first batch of comments.

PTAL

alan-kut commented on network/tools/network-policy-enforcement-latency/pkg/client.go:
```diff
@@ -0,0 +1,552 @@
+		}
+	}(c.metricsServer, context.TODO())
+
+	if c.config.podCreationMode {
+		if err := c.measurePodCreation(); err != nil {
+			klog.Errorf("Pod creation test failed, error: %v", err)
+		}
+	} else {
+		if err := c.measureNetPolicyCreation(); err != nil {
+			klog.Errorf("Pod creation test failed, error: %v", err)
+		}
+	}
```
I don't like it. If I'm not mistake majority of the code is for either pod or policy.
Some common code could be extracted to some utils or library.

dlapcevic replied:
You are right. Most of the functionality besides initialization and sending a request is not common to the modes.
I will extract the common code to utils package and separate it into 2 applications. We can run both in the scale tests.
Sounds good?

alan-kut commented on network/tools/network-policy-enforcement-latency/pkg/client.go:
```diff
@@ -0,0 +1,552 @@
+		}
+
+		// Do the measurements only if it hasn't already been done for this pod.
+		if !targetUsed {
```
if targetUsed {
  return
}

dlapcevic replied:
Done

alan-kut commented on network/tools/network-policy-enforcement-latency/pkg/client.go:
```diff
@@ -0,0 +1,552 @@
+		return
+	}
+
+	if len(pod.Status.PodIP) > 0 {
```
if len(...) == 0 {
  return
}

dlapcevic replied:
Done

alan-kut commented on network/tools/network-policy-enforcement-latency/pkg/client.go:
```diff
@@ -0,0 +1,552 @@
+				namespace: pod.GetNamespace(),
+				startTime: startTime,
+			}
+			go c.recordFirstSuccessfulRequest(target, true)
```
You don't want to wait for it?

dlapcevic replied:
No. It contains the functionality of recording the first successful request, by generating metrics.

alan-kut commented on network/tools/network-policy-enforcement-latency/pkg/client.go:
```diff
@@ -0,0 +1,552 @@
+	}
+
+	c.policyCreatedTime.lock.Unlock()
+	if failed {
```
What happens in this case?
Don't we lose some test data because of this?
Doesn't it mean that request was sent before the policy was deployed?
Can't we wait for the policy before testing reachability?

dlapcevic replied:
Policy creation test starts sending traffic before allow policy is created.
Once allow policy is created, the requests to pods will succeed, and the policy will exist to measure latency for its creation.
If the policy cannot be retrieved here, it means the setup is incorrect. The warning log will make it clear that the specified allow policy is not found. No metrics should be generated from it.

dlapcevic commented:
I refactored the code. Extracted all common code into separate package and split the modes into two applications `pod-creation-reachability-latency` and `policy-creation-enforcement-latency`.
@alan-kut Please take another look.

dlapcevic commented:
Hey @marseel, can you give LGTM for this one, so I can merge it?
@alan-kut reviewed it.

marseel commented:
/lgtm
